packages/react-dom/src/client/ReactDOM.js

```diff
@@ -759,9 +766,90 @@ function createPortal(
   return ReactPortal.createPortal(children, container, null, key);
 }
 
+type WorkNode = FiberRootWorkNode & {
+  then(onComplete: () => mixed): void,
+  commit(): void,
+
+  _reactRootContainer: FiberRoot,
+
+  _completionCallbacks: Array<() => mixed> | null,
+  _didComplete: boolean,
+};
+
+function insertWork(root: FiberRoot, work: FiberRootWorkNode) {
+  // Insert into root's list of work nodes.
+  const expirationTime = work._expirationTime;
+  const firstNode = root.firstTopLevelWork;
+  if (firstNode === null) {
+    root.firstTopLevelWork = work;
+    work._next = null;
+  } else {
+    // Insert sorted by expiration time then insertion order
+    let insertAfter = null;
+    let insertBefore = firstNode;
+    while (
+      insertBefore !== null &&
+      insertBefore._expirationTime <= expirationTime
+    ) {
+      insertAfter = insertBefore;
+      insertBefore = insertBefore._next;
+    }
+    work._next = insertBefore;
+    if (insertAfter !== null) {
+      insertAfter._next = work;
+    }
+  }
+}
+
+function Work(root: FiberRoot, defer: boolean, expirationTime: ExpirationTime) {
+  this._reactRootContainer = root;
+
+  this._defer = defer;
+  this._expirationTime = expirationTime;
+  this._next = null;
+
+  this._completionCallbacks = null;
+  this._didComplete = false;
+}
+Work.prototype._onComplete = function() {
+  this._didComplete = true;
+  // Set this to null so it isn't called again.
+  this._onComplete = null;
+  const callbacks = this._completionCallbacks;
+  if (callbacks === null) {
+    return;
+  }
+  for (let i = 0; i < callbacks.length; i++) {
+    // TODO: Error handling
```

Comment:
TODO for when?

Reply:
For whenever we figure out how errors inside a `then` callback should be handled :D Would like to do that before landing this PR, if we intend this to ship in 16.1.

Reply:
My current thinking is that `then` callbacks should always resolve when the tree completes, regardless of errors. I don't think we can/should support rejection handlers in the case of unhandled render errors, because it won't be clear which update caused the error. Could be from a previous or subsequent update.

What I'm not sure about is the rethrow behavior. If there's an unhandled render error, and the `then` callback throws, which error should we rethrow at the end?